### PR TITLE
SQL: Ibis frontend

### PR DESCRIPTION
**Pull request by webmiche**
This patch adds a first version of an ibis frontend. The frontend currently supports exactly the query `table.filter(table['a'] = 'AS').

The frontend consists of a basic visitorpattern with a `visit()`-function, that matches on the given node and dispatches to the proper function to build the operation for that node.

### CODE REVIEW

File: experimental/sql/tools/rel_opt.py [hunk 1]
```diff
@@ -4,12 +4,33 @@
...
+from src.ibis_frontend import ibis_to_xdsl
 from dialects.ibis_dialect import Ibis
 
 
 class RelOptMain(xDSLOptMain):
 
+  def register_all_frontends(self):
+    super().register_all_frontends()
+
+    def parse_ibis(f: IOBase):
```
Comment (webmiche) [inline experimental/sql/tools/rel_opt.py:18]: This is still something I am not very proud of and it comes from the fact that this rel_opt tool wants to read one file and build an IR from that. I think it might make sense to extensively test using `pytest` further down the line (AFAIU, the same framework ibis uses for their test-suite).

Comment (ingomueller-net) [reply experimental/sql/tools/rel_opt.py:18]: How about a while loop on `f.readline()` that filters out lines that `startswith(#)`? Similar to [this SO post](https://stackoverflow.com/a/6985913/651937) and potentially as a [generator](https://stackoverflow.com/a/519653/651937).

Comment (ingomueller-net) [reply experimental/sql/tools/rel_opt.py:18]: Actually, if you prefix the `CHECK*` with ` #`, `eval`ing the whole file could be fine?

**Review by ingomueller-net — COMMENTED** [review]
LGTM modulo minor comments.

File: experimental/sql/tools/rel_opt.py (hunk 1 above)
Comment (webmiche) [reply experimental/sql/tools/rel_opt.py:18]: Still, I don't like this approach a lot, since the query actually isn't generated form just the file, but from some local context within rel_opt as well...

File: experimental/sql/src/ibis_frontend.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import ibis.expr.operations.generic as gen_types
+from ibis.expr.operations.logical import Equals as EQ
+import ibis.backends.pandas.client as PandasBackend
+
+import dialects.ibis_dialect as id
+
+# This file contains the translation from ibis to the ibis_dialect.
+# It is built in a visitor-style, as this is one of the best designs
+# to crawl trees in my experience. The entry-point is the `visit`-
+# function just below. It dispatches to the appropriate translator.
```
[the remark this answers is not in the document]
Comment (webmiche) [reply experimental/sql/src/ibis_frontend.py]: I wrote something on that.

File: experimental/sql/src/ibis_frontend.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+
+@dataclass
+class NodeVisitor:
```
[the remark this answers is not in the document]
Comment (webmiche) [reply experimental/sql/src/ibis_frontend.py]: I rewrote everything using this package. It feels way nicer to me, particularly since the top-level visit, that just calls the others, is not needed anymore. Plus, I like shifting as much complexity to others as possible.

File: experimental/sql/src/ibis_frontend.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    if isinstance(op, gen_types.TableColumn):
+      table = Region.from_operation_list([self.visit(op.table)])
+      new_op = id.TableColumn.get(table, op.name)
+      return new_op
+    raise Exception(f"Unknown stringcolumn: {type(op)}")
+
+  def visit_BooleanColumn(
+      self, boolColumn: ibis.expr.types.BooleanColumn) -> Operation:
+    op = boolColumn.op()
+    if isinstance(op, gen_types.TableColumn):
```
[the remark this answers is not in the document]
Comment (webmiche) [reply experimental/sql/src/ibis_frontend.py]: I think this should be handled by the fact that we have a @dispatch(ibis.expr.types.BooleanColumn) just before.

File: experimental/sql/src/ibis_frontend.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  def visit_TableExpr(self, table: ibis.expr.types.TableExpr) -> Operation:
+    op = table.op()
+    if isinstance(op, PandasBackend.PandasTable):
+      schema = self.visit_Schema(op.schema)
+      new_op = id.PandasTable.get(op.name, schema)
+      return new_op
+    if isinstance(op, rels.Selection):
+      table = Region.from_operation_list([self.visit(op.table)])
+      # TODO: handle mulitple predicates
```
[the remark this answers is not in the document]
Comment (webmiche) [reply experimental/sql/src/ibis_frontend.py]: Yes indeed, but we will still have to think about how this will be represented in the IR and how ibis represents this currently. The most straightforward way I see is just having multiple operations in the region, each defining a condition, and the actual predicate is the `and` of all of those. But how do we then represent `or`s?

This will probably be heavily influenced by ibis for this layer, but it is certainly also a question for lower layers.

File: experimental/sql/test/frontend/selection.ibis [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# RUN: rel_opt.py -f ibis %s | filecheck %s
+
+table.filter(table['a'] == 'AS')
+
+# CHECK: ibis.selection() {
```
Comment (nicolasvasilache) [inline experimental/sql/test/frontend/selection.ibis:5]: nit: can you add a few whitespace betwee `#` and `CHECK` to align the `:` ?

File: experimental/sql/tools/rel_opt.py (hunk 1 above)
Comment (nicolasvasilache) [reply experimental/sql/tools/rel_opt.py:18]: I am not sure I follow 100% but maybe you could add a fake test_harness op that would take l19-28 as a python string and interpret it on the fly.

**Review by nicolasvasilache — APPROVED** [review]

File: experimental/sql/tools/rel_opt.py (hunk 1 above)
Comment (ingomueller-net) [reply experimental/sql/tools/rel_opt.py:18]: Maybe the context should be part of the test case? It could still be shortened slightly by moving some of it to some `test_utils` package or similar.

```python
import ibis
import pandas as pd

connection = ibis.pandas.connect(
  {"t": pd.DataFrame({"a": ["AS", "EU", "NA"]})})
table = connection.table('t')
```

But maybe that doesn't combine well with `eval`? I think that the root cause of this whole issue is the fact that Ibis isn't a language but an API...

Eventually, I think that we want to call `ibis_to_xdsl` from our own Ibis backend. But writing that could involve a serious amount of plumbing, so I wouldn't do it right now.

I am inclined to say to kick the can down the road: live it as is for now and worry about it later.

File: experimental/sql/src/ibis_frontend.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ingomueller-net) [reply experimental/sql/src/ibis_frontend.py]: Cool, yes, that looks better.

Question: is it such that *every* expression has *exactly one* `op`? Would it make sense to visit any expression type with the same visitor, take the op out, and then visit that op (which dispatches to op-specific implementations)?

File: experimental/sql/tools/rel_opt.py (hunk 1 above)
Comment (webmiche) [reply experimental/sql/tools/rel_opt.py:18]: Let me rephrase: Fundamentally, I feel that we are not really generating IR from a file like a *file*check test usually does, but rather from a file and a context. The exact schema of a table for example is part of that context. A linked part would be where to find column "a" in a tuple of table "t" (at position 0? or 1?) or where "t" actually is in memory. This is information that we will need at some point to actually be able to lower, but I am not sure whether this should be part of the IR of a query.

I would suggest leaving it as is and reopening this discussion once we start lowering to less abstract representations. Then we might be able to see more easily if this information should be part of the IR, or if this should be passed into the lowering pass as a context argument.

File: experimental/sql/src/ibis_frontend.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ingomueller-net) [reply experimental/sql/src/ibis_frontend.py]: What I was thinking is that there is an `AndExpression`, such that `$c > 1 and $b == "AS"` is translated into `AndExpression(Lt(Col("c"), Lit(1)), Eq(Col("b"), Lit("AS")))`. More predicates can then be expressed as nested `AndExpression`s but there is always just a single root expression.

File: experimental/sql/tools/rel_opt.py (hunk 1 above)
Comment (ingomueller-net) [reply experimental/sql/tools/rel_opt.py:18]: OK, I think I understand. And I agree: let's leave it as is for now.

Comment (nicolasvasilache) [reply experimental/sql/tools/rel_opt.py:18]: sgtm!
